Machine ID Docs Refactor: Application Access (#32745)
gravitational/teleport: docs/pages/machine-id/access-guides/applications.mdx

---
title: Machine ID with Application Access
description: How to use Machine ID to access applications
---

(!docs/pages/includes/machine-id/v2-config-warning.mdx!)

> Teleport Application Access protects and controls access to HTTP and TCP

Reviewer: ```suggestion
Teleport protects and controls access to HTTP and TCP
```
So "Application Access" doesn't seem like a distinct product (we're not positioning it as one anymore).

Author: Wasn't aware of this change - thank you.

applications. Machine ID can be used to grant machines secure, short-lived
access to these applications.

In this guide, you will configure `tbot` to produce credentials that can be

> used to access an application configured in Teleport Application Access.

Reviewer: ```suggestion
used to access an application enrolled in your Teleport cluster.
```

## Prerequisites

(!docs/pages/includes/edition-prereqs-tabs.mdx!)

- If you have not already connected your application to Teleport, follow
  the [Application Access Getting Started Guide](../../application-access/getting-started.mdx).
- (!docs/pages/includes/tctl.mdx!)
- `tbot` must already be installed and configured on the machine that will
  access applications. For more information, see the
  [platform guides](../platform-guides.mdx).

## Step 1/3. Configure RBAC

> First, Teleport must be configured to allow the credentials produced by the bot

Reviewer: Same comment as [this one](https://github.com/gravitational/teleport/pull/32741/files#r1344546883) in the Ansible guide

Author: 65205ceaea

to access an Application.

If you have followed an access guide, you will have created a role and granted
the bot the ability to impersonate it already. This role just needs to have the
additional privileges added to it.

Use `tctl edit role/example-bot` to add the following rule to the role:

```yaml
spec:
  allow:
    # Grants access to all applications.
    app_labels:
      '*': '*'
```

This grants access to all applications. In production environments you should
modify these labels to grant access to only the applications that the machine
will need access to.

## Step 2/3. Configure an application `tbot` output

> Now, `tbot` needs to be configured with an output that to produce the

Reviewer: ```suggestion
Now, `tbot` needs to be configured with an output to produce the
```

> credentials needed for Application Access. To do this, the `application` output

Reviewer: ```suggestion
credentials needed to access applications in your infrastructure. To do this, the `application` output
```

type is used.

The application you want the credentials to have access to must be specified
using the `app_name` field. In this example, the debug application (`dumper`)
will be used.

Outputs must be configured with a destination. In this example, the `directory`
destination will be used. This will write artifacts to a specified directory on
disk. Ensure that this directory can be written to by the Linux user that
`tbot` runs as, and that it can be read by the Linux user that will be accessing
applications.

Modify your `tbot` configuration to add an `application` output:

```yaml
outputs:
- type: application
  # specify the name of the application you wish the credentials to grant
  # access to.
  app_name: dumper
  destination:
    type: directory
    # For this guide, /opt/machine-id is used as the destination directory.
    # You may wish to customize this. Multiple outputs cannot share the same
    # destination.
    path: /opt/machine-id
```

> Ensure you replace `grafana` with the name of the application as registered in

Reviewer: ```suggestion
Ensure you replace `dumper` with the name of the application you registered in
```

Teleport.

If operating `tbot` as a background service, restart it. If running `tbot` in
one-shot mode, it must be executed before you attempt to use the credentials.

## Step 3/3. Connect to your web application with the Machine ID identity

Once `tbot` has been run, credentials will be output to the directory specified
in the destination. Using the example of `/opt/machine-id`:

- `/opt/machine-id/tlscert`: the client TLS certificate
- `/opt/machine-id/key`: the TLS certificate's private key

You may use these credentials with any client application that supports them.

The Teleport Proxy makes apps available via subdomains of its public web
address. Given the debug application named `dumper` and a Teleport Proxy at
`https://example.teleport.sh:443`, the app may be accessed at
`https://dumper.example.teleport.sh:443`.

> For example, to access the application using `curl`:

Reviewer: ```suggestion
For example, to access the application using `curl`:

```

Note that if the certificates are invalid or otherwise misconfigured, clients
will be redirected to the Teleport login page when attempting to access the app.

### Authenticated tunnel

For cases where the client you wish to use to connect to the application does not
support client certificates, it is possible to open an authenticated tunnel.

> This will listen on a local port and automatically attach the credentials

Reviewer: ```suggestion
This will listen on a local port and automatically attach the credentials,
```
